File: GameClient/Assets/Scripts/GameLogic/Login/LoginController.cs
using System.Collections;
using System.Collections.Generic;
using UnityEngine;
using Common.Protocol;
using Common;
using System;

public class LoginController : Singleton<LoginController>
{
    LoginView view;
    //UserData userData;
    //LoginModel model;

    public LoginController()
    {
        view = new LoginView();
        //model = new LoginModel(); 
        //userData = new UserData();
    }

    public void RegisterProto()
    {
        NetworkManager.Instance.RegisterProto((int)ProtocolId.ReturnUserDataProtocol, Debuglog);
        NetworkManager.Instance.RegisterProto((int)ProtocolId.ReturnUserDataProtocol, Verify);
    }

    public void UnregisterProto()
    {
        NetworkManager.Instance.UnregisterProto((int)ProtocolId.ReturnUserDataProtocol, Debuglog);
        NetworkManager.Instance.UnregisterProto((int)ProtocolId.ReturnUserDataProtocol, Verify);
    }

    public void Debuglog(object param)
    {
        Debug.Log("login");
    }

    public void Verify(object param)
    {
        byte[] data = (byte[])param;
        ReturnUserDataProtocol returnUserDataProtocol = new ReturnUserDataProtocol();
        returnUserDataProtocol.Decode(data);
        if (returnUserDataProtocol.Result)
        {
            //1，save userData
            //2,进入GameHall
            UserData userData = new UserData(returnUserDataProtocol.Id, returnUserDataProtocol.Username);
            PlayerInfoManager.Instance.SetUserData(userData);
            StateMachine.Instance.ChangeState(StateType.GameHall);
        }
        else
        {
            Debug.Log("fail");
        }
    }   

    public void Login(string username, string pwd)
    {
        //1,set Data to _model
        //2,NetManager发送请求       

        NetworkManager.Instance.Send(new LoginProtocol(username, pwd));
    }

    public void ShowView()
    {
        UIManager.Instance.AddView(view);        
    }

    public void CloseView()
    {
        UIManager.Instance.RemoveView(view);
    }


    
}

File: GameClient/Assets/HotFix/Lua/MVC/Login/LoginController.lua
local loginController = class()

function loginController:ctor(  )
	self.loginView = require("MVC/Login/LoginView"):new()
end

function loginController:registerProto(  )
	NetworkManager.Instance:RegisterProto(ProtocolId.ReturnUserDataProtocol,packFunction(self,self.verify))
end

function loginController:unregisterProto(  )
	NetworkManager.Instance:UnregisterProto(ProtocolId.ReturnUserDataProtocol,packFunction(self,self.verify))
end

function loginController:login(  )
	NetworkManager.Instance:Send(Common.Protocol.LoginProtocol(self.loginView.username, self.loginView.pwd))
end

function loginController:showView(  )
	global.UIManager:addView(self.loginView)
end

function loginController:closeView(  )
	global.UIManager:removeView(self.loginView)
end

function loginController:verify( param )
	local returnUserData = Common.Protocol.ReturnUserDataProtocol()
	returnUserData:Decode(param)

	if returnUserData.Result then
		userData.id =  returnUserData.Id
		userData.username =	returnUserData.Username
		global.FSM:changeState("FSM/GameHallState")
	end

end

return  loginController
